balancer/pickfirstleaf/pickfirstleaf.go

```diff
@@ -0,0 +1,647 @@
+		serializerCancel: cancel,
+		state:            connectivity.Connecting,
+	}
+	b.logger = internalgrpclog.NewPrefixLogger(logger, fmt.Sprintf(logPrefix, b))
+	return b
+}
+
+func (b pickfirstBuilder) Name() string {
+	return Name
+}
+
+func (pickfirstBuilder) ParseConfig(js json.RawMessage) (serviceconfig.LoadBalancingConfig, error) {
+	var cfg pfConfig
+	if err := json.Unmarshal(js, &cfg); err != nil {
+		return nil, fmt.Errorf("pickfirst: unable to unmarshal LB policy config: %s, error: %v", string(js), err)
+	}
+	return cfg, nil
+}
+
+type pfConfig struct {
+	serviceconfig.LoadBalancingConfig `json:"-"`
+
+	// If set to true, instructs the LB policy to shuffle the order of the list
+	// of endpoints received from the name resolver before attempting to
+	// connect to them.
+	ShuffleAddressList bool `json:"shuffleAddressList"`
+}
+
+// scData keeps track of the current state of the subConn.
+type scData struct {
+	// The following fields are initialized at build time and read-only after
+	// that.
+	subConn balancer.SubConn
+	addr    resolver.Address
+
+	// The following fields should only be accessed from a serializer callback
+	// to ensure synchronization.
+	state   connectivity.State
+	lastErr error
+}
+
+func (b *pickfirstBalancer) newSCData(addr resolver.Address) (*scData, error) {
+	sd := &scData{
+		state: connectivity.Idle,
+		addr:  addr,
+	}
+	sc, err := b.cc.NewSubConn([]resolver.Address{addr}, balancer.NewSubConnOptions{
+		StateListener: func(state balancer.SubConnState) {
+			// Store the state and delegate.
+			b.serializer.TrySchedule(func(_ context.Context) {
+				sd.state = state.ConnectivityState
+				b.updateSubConnState(sd, state)
+			})
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	sd.subConn = sc
+	return sd, nil
+}
+
+type pickfirstBalancer struct {
+	// The following fields are initialized at build time and read-only after
+	// that and therefore do not need to be guarded by a mutex.
+	logger *internalgrpclog.PrefixLogger
+	cc     balancer.ClientConn
+
+	// The serializer and its cancel func are initialized at build time, and the
+	// rest of the fields here are only accessed from serializer callbacks (or
+	// from balancer.Balancer methods, which themselves are guaranteed to be
+	// mutually exclusive) and hence do not need to be guarded by a mutex.
+	// The serializer is used to ensure synchronization of updates triggered
+	// from the idle picker and the already serialized resolver,
+	// subconn state updates.
+	serializer       *grpcsync.CallbackSerializer
+	serializerCancel func()
+	state            connectivity.State
+	subConns         *resolver.AddressMap // scData for active subonns mapped by address.
+	addressList      addressList
+	firstPass        bool
+}
+
+func (b *pickfirstBalancer) ResolverError(err error) {
+	doneCh := make(chan error, 1)
+	b.serializer.ScheduleOr(func(_ context.Context) {
+		b.resolverError(err)
+		close(doneCh)
+	}, func() {
+		close(doneCh)
+	})
```

Comment:
Don't you have to block on `doneCh` at the end of this method?

Reply:
Oops, added the channel wait.